geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/SlotAdvisor.java

```diff
@@ -0,0 +1,207 @@
+    try {
+      RedisMemberInfo memberInfo = getMemberInfo(key.getBucketId());
+      return Pair.of(memberInfo.getHostAddress(), memberInfo.getRedisPort());
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(ex);
+    }
+  }
+
+  public Map<String, List<Integer>> getMemberBuckets() {
+    Map<String, List<Integer>> memberBuckets = new HashMap<>();
+    for (int bucketId = 0; bucketId < REDIS_REGION_BUCKETS; bucketId++) {
+      String memberId = getOrCreateMember(bucketId).getUniqueId();
+      memberBuckets.computeIfAbsent(memberId, k -> new ArrayList<>()).add(bucketId);
+    }
+
+    return memberBuckets;
+  }
+
+  /**
+   * This returns a list of {@link MemberBucketSlot}s where each entry corresponds to a bucket. If
+   * the details for a given bucket cannot be determined, that entry will contain {@code null}.
+   */
+  public synchronized List<MemberBucketSlot> getBucketSlots() {
+    List<MemberBucketSlot> memberBucketSlots = new ArrayList<>(RegionProvider.REDIS_REGION_BUCKETS);
+    try {
+      for (int bucketId = 0; bucketId < REDIS_REGION_BUCKETS; bucketId++) {
+        RedisMemberInfo memberInfo = getMemberInfo(bucketId);
+        if (memberInfo != null) {
+          memberBucketSlots.add(
+              new MemberBucketSlot(bucketId, memberInfo.getHostAddress(),
+                  memberInfo.getRedisPort()));
+        }
+      }
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(ex);
+    }
+
+    return memberBucketSlots;
+  }
+
+  private InternalDistributedMember getOrCreateMember(int bucketId) {
+    return dataRegion.getOrCreateNodeForBucketWrite(bucketId, null);
+  }
+
+  /**
+   * This method will retry for {@link #HOSTPORT_RETRIEVAL_ATTEMPTS} attempts and return null if
+   * no information could be retrieved.
+   */
+  private RedisMemberInfo getMemberInfo(int bucketId) throws InterruptedException {
+    RedisMemberInfo response;
+
+    for (int i = 0; i < HOSTPORT_RETRIEVAL_ATTEMPTS; i++) {
+      response = getMemberInfo0(bucketId);
+      if (response != null) {
+        return response;
+      }
+
+      Thread.sleep(HOSTPORT_RETRIEVAL_INTERVAL);
+    }
+
+    logger.error("Unable to retrieve host and redis port for member with bucketId: {}", bucketId);
+
+    return null;
+  }
+
+  @SuppressWarnings("unchecked")
+  private RedisMemberInfo getMemberInfo0(int bucketId) {
+    InternalDistributedMember member = getOrCreateMember(bucketId);
+
+    if (memberInfos.containsKey(member)) {
+      return memberInfos.get(member);
+    }
+
+    List<RedisMemberInfo> memberInfos;
+    try {
+      ResultCollector<RedisMemberInfo, List<RedisMemberInfo>> resultCollector =
+          FunctionService.onRegion(dataRegion).execute(RedisMemberInfoRetrievalFunction.ID);
+      memberInfos = resultCollector.getResult();
+    } catch (Exception e) {
```

Comment:
Since you've changed the function to be registered earlier, I don't think it makes sense to catch and ignore function failures anymore.